spec.bs

```diff
@@ -289,46 +322,179 @@ To <dfn for="PrivateAggregation">obtain an aggregatable report</dfn> given an [=
+To <dfn>attempt to queue reports for sending</dfn> given a [=list=] of [=aggregatable reports=] |reports|:
+1. [=list/iterate|For each=] |report| of |reports|, run these steps [=in parallel=]:
+    1. Run these steps, but [=abort when=] the [=user agent=] shuts down:
+        1. If |report|'s [=aggregatable report/queued=] value is true, return.
+        1. Set |report|'s [=aggregatable report/queued=] value to true.
+        1. If |report|'s [=aggregatable report/report time=] is less than the [=wall clock=]'s [=current unsafe time=], set |report|'s [=aggregatable report/report time=] to [=wall clock=]'s [=current unsafe time=] plus an [=implementation-defined=] random non-negative [=duration=].
 
-Issue: Don't we need to take into account the time "now" as well?
+            Note: On startup, it is possible the user agent will need to send many reports whose report times passed while the browser was closed. Adding random delay prevents temporal joining of reports.
+        1. Wait until the [=wall clock=]'s [=current unsafe time=] is equal to |report|'s [=aggregatable report/report time=].
+        1. Optionally, wait a further [=implementation-defined=] non-negative [=duration=].
 
-Sending reports {#sending-reports}
-----------------------------------
+            Note: This is intended to allow user agents to optimize device resource usage.
+        1. Run [=attempt to deliver a report=] with |report|.
+    1. [=If aborted=], set |report|'s [=aggregatable report/queued=] value to false.
+
+        Note: It might be more practical to perform these steps when the [=user agent=] next starts up.
+
+To <dfn>attempt to deliver a report</dfn> given an [=aggregatable report=] |report|:
+1. Let |url| be the result of [=obtaining a reporting endpoint=] given |report|'s [=aggregatable report/reporting origin=] and |report|'s [=aggregatable report/api=].
+1. Let |data| be the result of [=serializing an aggregatable report=] given |report|.
+1. If |data| is an error, [=list/remove=] |report| from the [=aggregatable report cache=]. (TODO: do we need to queue this task?)
+1. Let |request| be the result of [=creating a report request=] given |url| and |data|.
+1. [=Queue a task=] to [=fetch=] |request| with [=fetch/processResponse=] being the following steps that take a [=response=] |response|:
+    1. Let |shouldRetry| be false.
+    1. If |response|'s [=response/status=] is not 200, optionally set |shouldRetry| to true.
```

Comment:
Do we need to call this out explicitly, or just say in case of error?

Reply:
I think I'd lean towards either leaving it as is or making |shouldRetry| completely implementation-defined. (Just not sure "error" by itself is clearly defined here.)

Reply:
Maybe just make |shouldRetry| implementation-defined to avoid ambiguity?

Reply:
Went with that, but added a "should" sentence -- I think it not being a requirement makes me more okay with the definition of an error being potentially uncertain, but lmk what you think :)